Soft body

AndreaCatania opened the pull request:
This is the implementation of soft body, not completed at 100% but still work.

Minimal test project: 
[platformer.zip](https://github.com/godotengine/godot/files/1444331/platformer.zip)

reduz commented:
this seems ok, i need to add some functions to VisualServer so the mesh can
get updated..

On Sun, Nov 5, 2017 at 4:40 PM, Andrea Catania <notifications@github.com>
wrote:

> This is the implementation of soft body, not completed at 100% but still
> work.
>
> Minimal test project:
> platformer.zip
> <https://github.com/godotengine/godot/files/1444331/platformer.zip>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/godotengine/godot/pull/12678
> Commit Summary
>
>    - Created class to support soft body
>
> File Changes
>
>    - *M* core/math/triangle_mesh.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-0> (30)
>    - *M* core/math/triangle_mesh.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-1> (1)
>    - *A* editor/icons/icon_soft_physics_body.svg
>    <https://github.com/godotengine/godot/pull/12678/files#diff-2> (56)
>    - *M* editor/plugins/mesh_instance_editor_plugin.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-3> (32)
>    - *M* editor/plugins/mesh_instance_editor_plugin.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-4> (2)
>    - *M* editor/spatial_editor_gizmos.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-5> (131)
>    - *M* editor/spatial_editor_gizmos.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-6> (23)
>    - *M* modules/bullet/bullet_physics_server.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-7> (177)
>    - *M* modules/bullet/bullet_physics_server.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-8> (40)
>    - *M* modules/bullet/soft_body_bullet.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-9> (250)
>    - *M* modules/bullet/soft_body_bullet.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-10> (86)
>    - *M* scene/3d/area.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-11> (2)
>    - *M* scene/3d/collision_object.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-12> (127)
>    - *M* scene/3d/collision_object.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-13> (14)
>    - *M* scene/3d/collision_shape.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-14> (2)
>    - *M* scene/3d/physics_body.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-15> (22)
>    - *M* scene/3d/physics_body.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-16> (1)
>    - *A* scene/3d/soft_physics_body.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-17> (615)
>    - *A* scene/3d/soft_physics_body.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-18> (149)
>    - *M* scene/register_scene_types.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-19> (6)
>    - *M* scene/resources/mesh.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-20> (31)
>    - *M* scene/resources/mesh.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-21> (5)
>    - *A* scene/resources/soft_body_shape.cpp
>    <https://github.com/godotengine/godot/pull/12678/files#diff-22> (208)
>    - *A* scene/resources/soft_body_shape.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-23> (110)
>    - *M* servers/physics/physics_server_sw.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-24> (66)
>    - *M* servers/physics_server.h
>    <https://github.com/godotengine/godot/pull/12678/files#diff-25> (66)
>
> Patch Links:
>
>    - https://github.com/godotengine/godot/pull/12678.patch
>    - https://github.com/godotengine/godot/pull/12678.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/12678>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2_zNwXF2WviUAuR-SvJ9S1se2Bsdks5szg9EgaJpZM4QSgC1>
> .
>

kubecz3k commented:
@AndreaCatania There is some problem with dependencies in the demo + https://i.imgur.com/WQsH3xO.png

AndreaCatania commented:
@kubecz3k are you sure that you checked the project with the correct binary? Because the previous commit had a compilation problem in linux.

Check it now

reduz commented:
Would appreciate squashing, so its easier to read

AndreaCatania commented:
@reduz let me complete it I could need the history. I'm quite near the end

reduz commented:
ah, alright

On Fri, Nov 17, 2017 at 9:42 PM, Andrea Catania <notifications@github.com>
wrote:

> @reduz <https://github.com/reduz> let me complete it I could need the
> history. I'm quite near the end
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/12678#issuecomment-345403026>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2-5QYQSUbGPhCsDJbJluAFOOEJldks5s3if9gaJpZM4QSgC1>
> .
>

AndreaCatania commented:
This is a video to show how looks soft body in this moment: https://www.youtube.com/watch?v=8A6PYFSesKQ&feature=youtu.be

Chaosus commented:
This PR delayed to 3.1 ? So sad...

reduz commented:
We are pretty close with it, and it does not affect other functions, so we
may merge it before 3.0


On Mon, Nov 20, 2017 at 10:25 AM, Yuri Roubinski <notifications@github.com>
wrote:

> This PR delayed to 3.1 ? So sad...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/12678#issuecomment-345694344>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z29IUZeVVZSfQrq9gKuN6P8PRBJFbks5s4X2ugaJpZM4QSgC1>
> .
>

AndreaCatania committed:
Added function append_array to vector

AndreaCatania committed:
Added some API to visual server so from control VRAM buffer is more easy

AndreaCatania committed:
Implemented Soft body
- Soft Body Physics node
- Soft Body Rendering
- Soft body Editor
- Soft body importer

AndreaCatania commented:
I've fixed last problems and squashed all commits. Now it should be almost completed.

I've to expose last properties (that are some advanced features of SoftBody)

@reduz Now you can check the code easily.

AndreaCatania commented:
One last thing, If you want to test it remember to active the softbody in the settings

_physics > 3d_

AndreaCatania commented:
I've tested how implement clusters in order to allow Soft vs Soft collision but it doesn't works well right now.

http://bulletphysics.org/Bullet/phpBB3/viewtopic.php?f=9&t=8803

So the implementation of softbody is completed.

akien-mga commented:
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

AndreaCatania commented:
Sorry but I need to write down a note :)

> NOTE:
> Reduz advised me to create a buffer where the surface information are decompressed, so the physics server will have the possibility to update soft body node position directly.
In this way we will skip a memory allocation.
>
>  https://github.com/godotengine/godot/blob/master/servers/visual_server.cpp#L346
>  or here: https://github.com/godotengine/godot/blob/master/servers/visual_server.cpp#L798
>  this is where the buffers are created from the arrays.. in your case you should just write to a  buffer directly (locking a mutex) and when visual server calls that signal (which may be from another thread) you lock again and call the update region function

karroffel commented:
Oh the github codeowner thing is pretty cool :D

The GLES2 code should pretty much be the same as for GLES3, but you can leave it out and I fill it in

AndreaCatania commented:
@karroffel yep it's really cool :D 

Also yes I prefer that you do it when you'll implement other things

AndreaCatania commented:
I've rebased it with last engine changes and now it compiles correctly so this PR is ready to be merged, if no more things to change.

reduz commented on drivers/dummy/rasterizer_dummy.h:
```diff
@@ -322,6 +322,8 @@ class RasterizerStorageDummy : public RasterizerStorage {
 		return m->blend_shape_mode;
 	}
 
+	uint32_t mesh_surface_get_stride_in_array(RID p_mesh, int p_surface, int p_array_index) const { return 0; }
```
I think I would probably make this function go in VisualServer and not make it virtual, something like:

uint32_t mesh_surface_get_format_stride(int p_format) const;

so it s more or less generic and does not requeire reimplementing in every platform

reduz replied:
Also does not have the limitation of stalling the render thread if called.

AndreaCatania replied:
ok

reduz commented on scene/3d/collision_object.h:
```diff
@@ -38,7 +38,17 @@ class CollisionObject : public Spatial {
 
-	bool area;
+public:
+	enum CollisionObjectType {
```
I am still not convinced on this, given SoftBody does not use shapes. 

Also not convinced on modifying MeshInstance for this.

Does it not make more sense that SoftBody does everything as a single node and inherits GeometryInstance? This way there is no need to modify neither of these nodes to add softbody.

AndreaCatania replied:
The thing is that softbody use as collision shape the mesh that is rendered, but I agree with you this approach so good. Also inherit from GeometryInstance is not a good thing since softbody is a physical object mainly. A solution to this could be set the mesh resource directly to the soft body instead to use a MeshInstance, and let do all work inside the SoftBody

AndreaCatania replied:
But on the other hand the other problems are that I need to copy all functions from MeshInstance in order to allow artist to set Material and other things related to rendering..

AndreaCatania commented:
For some reason that project is not more usable.. But I can give you this one: 
[Soft.zip](https://github.com/godotengine/godot/files/2096752/Soft.zip)


Soft body friendly mean that the mesh is imported without compression and can be used to generate collision. There's an hint that say to import a mesh and select "SoftBody friendly" in the preset settings in the import panel

Supatier commented:
@AndreaCatania It works.

So I can't just create a new mesh from the editor and expect it to be "SoftBody friendly"? There's room for improvement.

AndreaCatania commented:
Yes I need to implement something like mesh converter that change how the mesh is stored inside the memory, I don't remember if there was some other reason why I didn't do it before.

reduz commented on editor/import/editor_import_collada.cpp:
```diff
@@ -89,10 +89,10 @@ struct ColladaImport {
 	Error _populate_skeleton(Skeleton *p_skeleton, Collada::Node *p_node, int &r_bone, int p_parent);
 	Error _create_scene_skeletons(Collada::Node *p_node);
 	Error _create_scene(Collada::Node *p_node, Spatial *p_parent);
-	Error _create_resources(Collada::Node *p_node, bool p_use_compression);
```
This huge amount of modifications to the importers are unnecessary, Simply work with `ResourceImporterScene::_fix_node()` and add a `-softbody` import flag. When this flag is found, you convert the mesh to a softbody friendly format.

reduz replied:
In fact, you can just convert it to a SoftBody node there.

AndreaCatania replied:
Ok

reduz commented on modules/bullet/space_bullet.cpp:
```diff
@@ -338,7 +339,7 @@ SpaceBullet::SpaceBullet(bool p_create_soft_world) :
 		contactDebugCount(0) {
 
-	create_empty_world(p_create_soft_world);
+	create_empty_world(GLOBAL_DEF("physics/3d/active_soft_world", false));
```
Is there any significant performance or memory implications of leaving it on by default?

AndreaCatania replied:
I didn't profiled it, when the soft world is not created all soft solvers are not set.

AndreaCatania commented:
@reduz Thanks you for reviews! I've removed all code to import softbody friendly mesh, because I've found the right way to set the necessary flags on the mesh array.

@Supatier Now is possible to simulate softbody even on mesh created in the editor!

@reduz Let me know you prefer to set this line: `GLOBAL_DEF("physics/3d/active_soft_world", false)` from false to true.

DarkEnchanter1 commented:
Attempting to create a softbody directly seems to cause a crash, so to make a softbody at the moment you need to create a meshinstance, attach a mesh, _then_ change it's type to softbody. I doubt this is intentional?

AndreaCatania commented:
@DarkEnchanter1 I've fixed it, thanks you for the report!
